### DISCUSSION TITLE
Question about CRUD addField value in Admin Panel

### DISCUSSION THREAD

**Discussion by wellyngtonlourenco** [Laravel-Backpack/community-forum#872] [Q&A (Help)]
Hello everyone!

I'm facing an issue in my admin panel setup, and I'd appreciate some guidance.

**Setup:**
- Laravel version: 10
- Backpack for Laravel version: 6 Free version

**Context:**
I have a `User` model that is related (`hasOne`) to two types (`BelongsTo`), namely `ClientInfo` and `EmployeeInfo`. For simplicity, I'll focus on `ClientInfo`. My goal is to have the update page for `ClientInfo` fill the fields with the related `User` data.

Here's a snippet of my code:

**User Model:**
```php
protected $fillable = [
    'name',
    'last_name',
    'email',
    'phone',
    'locale',
    'profile_picture',
    'password',
];

public function clientInfo(): \Illuminate\Database\Eloquent\Relations\HasOne
{
    return $this->hasOne(ClientInfo::class);
}
```

**ClientInfo Model:**
```php
protected $fillable = [
    'user_id'
];

public function user(): \Illuminate\Database\Eloquent\Relations\BelongsTo
{
    return $this->belongsTo(User::class);
}
```

**ClientInfoCrudController:**
```php
// List operation
$this->crud->addColumns([
    [
        'name'  => 'user.name',
        'label' => trans('backpack.label.user.name'),
        'type'  => 'text',
    ],
]}

// Create/Update fields
protected function setupCreateOperation()
{
    // ... (other setup)
    $this->addClientUserFields();
}

protected function setupUpdateOperation()
{
    $this->addClientUserFields();
}

protected function addClientUserFields(): void
{
    // Workaround to fill the text field with the correct user name
    // $user = $this->crud->getCurrentEntry() ? $this->crud->getCurrentEntry()->user : false;
    $this->crud->addFields([
        [
            'name'  => 'user.name',
            'label' => trans('backpack.label.user.name'),
            'type'  => 'text',
            // 'value'   => $user->name ?? '',
        ],
    ]);
}
```

**Question:**
Without my workaround, the ID of the `User` is displayed, so my workaround involves manually setting the value for the text field to avoid displaying the `User` ID. Is this approach correct? The goal is to achieve functionality similar to the `addColumns` setup, which correctly shows related user names.

Thank you for your help!

**Comment by pxpm** [comment]
Hello @wellyngtonlourenco 

I think that's due to the fact you are forcing the field to be a text field. In that case, since you are forcing it to "not be a relationship", it's just displaying the plain value "as text". 

You can either 
1 - remove the `type` from field definition.
2- use a closure in the `value` to process the raw value `function($entry) { return $entry->user->name; }`.

Let me know if that helps. 

Cheers

**Reply by wellyngtonlourenco** [reply]
Hello @pxpm!

Thanks for your answer!

1 - If i remove the `type`, the field become a select field with all the `User` Model entry names and the current entry one selected.

```
[
     'name'  => 'user.name',
     'label' => trans('backpack.label.user.name'),
     //'value' => function($entry) { return $entry->user->name; }
 ],
 [
     'name'  => 'user.last_name',
     'label' => trans('backpack.label.user.last_name'),
     'type'  => 'text',
     'value'   => $user->last_name ?? '',
 ],
```
<img width="224" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/54453599/a1780aef-2d5a-4ee2-9171-b078e0ee636b">


2 - This way also give me a select field but with no data preselected. 

```
 [
     'name'  => 'user.name',
     'label' => trans('backpack.label.user.name'),
     'value' => function($entry) { return $entry->user->name; }
 ],
 [
     'name'  => 'user.last_name',
     'label' => trans('backpack.label.user.last_name'),
     'type'  => 'text',
     //'value'   => $user->last_name ?? '',
 ],
```
<img width="390" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/54453599/315b5b17-206f-419a-b938-d4c9c24cd179">

**Reply by pxpm** [reply]
I am sorry, maybe I have missed the question. 

Let me try to clarify some things. What's the model you are using in this controller ? It is `ClientInfo` model right ? 

The `user()` relation on that model is a "belongsTo". Not an `hasOne` that you can edit on the same page using dot notation. 

What I mean, is that in your `UserCrudController` you can use `clientInfo.name` because `clientInfo` is a `hasOne` relation. 

In case of belongsTo, the most "similar" to create on the same page we have is `InlineCreate`. https://backpackforlaravel.com/docs/6.x/crud-fields#create-related-entries-in-a-modal-using-the-inlinecreate-operati

Let me know if that helps. 

Cheers

**Reply by wellyngtonlourenco** [reply]
Yes, this model is the `ClientInfo`.

For saving a new entry, i already figured out by hooking in the `store` method, so i just create both entries by myself.

The case i'm bringing here is about the Update page, when i edit a `ClientInfo` i also want to be able to edit the `User`. For updating the entry itself i'll do the same i did with the creation, my problem here is to output the `value` from the relationship. 

So when i'm setting up the update page, i want to pre-fill the user name for example from the relationship `BelongsTo` and i only could do that adding this variable:

```
$user = $this->crud->getCurrentEntry() ? $this->crud->getCurrentEntry()->user : false;
```
and then adding to the `value`:

```
[
     'name'  => 'user.name',
     'label' => trans('backpack.label.user.name'),
     'type'  => 'text',
     'value'   => $user->name ?? '',
 ]
```

So if there is no other way to access the value, no problem from my side i can keep like this, i just wanted to check because the `addColumn` works fine using the dot notation..

```
$this->crud->addColumns([
                                    [
                                        'name'  => 'user.name',
                                        'label' => trans('backpack.label.user.name'),
                                        'type'  => 'text',
                                    ],
```

Thanks for your time!

**Reply by pxpm** [reply]
Yes, indeed. addColumn works fine with dot notation because it uses `Arr:get()`, an helper that uses dot notation to access array data, and it eager loads the relationships defined in the columns.  

I would probably do it using the model or some repository for fields `$user = \App\MyModel::with('user')->find(request()->route()->parameter('id'))`

That way it would eager-load the user relation. 
Cheers.
